das/coordinator.go

```diff
@@ -0,0 +1,129 @@
+package das
+
+import (
+	"context"
+	"sync"
+
+	"github.com/celestiaorg/celestia-node/header"
+)
+
+// samplingCoordinator runs and coordinates sampling workers and updates current sampling state
+type samplingCoordinator struct {
+	concurrencyLimit int
+	getter           header.Getter
+	sampleFn         sampleFn
+
+	state coordinatorState
+
+	// resultCh fans-in sampling results from worker to coordinator
+	resultCh chan result
+	// updHeadCh signals to update network head header height
+	updHeadCh chan uint64
+	// waitCh signals to block coordinator for external access to state
+	waitCh chan func()
+
+	workersWg sync.WaitGroup
+	done
+}
+
+type result struct {
+	job
+	failed []uint64
+	err    error
+}
+
+func newSamplingCoordinator(
+	concurrency int,
+	samplingRange uint64,
+	getter header.Getter,
+	sample sampleFn) *samplingCoordinator {
+	return &samplingCoordinator{
+		concurrencyLimit: concurrency,
+		getter:           getter,
+		sampleFn:         sample,
+		state:            initState(samplingRange),
+		resultCh:         make(chan result),
+		updHeadCh:        make(chan uint64),
+		waitCh:           make(chan func()),
+		done:             newDone("sampling coordinator"),
+	}
+}
+
+func (sc *samplingCoordinator) run(ctx context.Context, cp checkpoint) {
+	sc.state.resumeFromCheckpoint(cp)
+	// resume workers
+	for _, wk := range cp.Workers {
+		sc.runWorker(ctx, sc.state.newJob(wk.From, wk.To))
+	}
+
+	for {
+		for !sc.concurrencyLimitReached() {
+			next, found := sc.state.nextJob()
+			if !found {
+				break
+			}
+			sc.runWorker(ctx, next)
+		}
+
+		select {
+		case max := <-sc.updHeadCh:
+			sc.state.updateHead(max)
+		case res := <-sc.resultCh:
+			sc.state.handleResult(res)
+		case wait := <-sc.waitCh:
+			wait()
+		case <-ctx.Done():
+			sc.workersWg.Wait()
+			sc.indicateDone()
+			return
+		}
+	}
+}
+
+// runWorker runs job in separate worker go-routine
+func (sc *samplingCoordinator) runWorker(ctx context.Context, j job) {
+	w := newWorker(j)
+	sc.state.putInProgress(j.id, w.getState)
+
+	// launch worker go-routine
+	sc.workersWg.Add(1)
+	go func() {
+		defer sc.workersWg.Done()
+		w.run(ctx, sc.getter, sc.sampleFn, sc.resultCh)
+	}()
+}
+
+// listen notifies coordinator about new header height has been found
+func (sc *samplingCoordinator) listen(ctx context.Context, height uint64) {
+	select {
+	case sc.updHeadCh <- height:
+	case <-ctx.Done():
+	}
+}
+
+// stats pause coordinator to get stats in concurrently safe manner
+func (sc *samplingCoordinator) stats(ctx context.Context) (SamplingStats, error) {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer wg.Done()
+
+	select {
+	case sc.waitCh <- wg.Wait:
```

Comment:
Thats another way of doing it. It could also be done with buffered channel, but implementation is more complicated. I feel like having a `Wait` is kinda helping to understand what is going on. I know this part is kinda confusing, so I've changed channel value type to full `*sync.WaitGroup` instead of  `func()` so there is less functional magic (but bigger interface passed).